phoenicis-repository/src/main/java/org/phoenicis/repository/types/GitRepository.java

```diff
@@ -65,72 +70,100 @@ private File createRepositoryLocation(String cacheDirectoryPath) {
 
-                        LOGGER.info("Pulling new commits from " + this);
+            // check that the repository folder exists
+            if (!folderExists) {
+                LOGGER.info("Creating local folder for " + this);
 
-                        gitRepository.pull().call();
-                    }
+                if (!this.localFolder.mkdirs()) {
+                    throw new RepositoryException("Couldn't create local folder for " + this);
+                }
+            }
+            Git gitRepository = null;
 
-                    result = localRepositoryFactory.createInstance(this.localFolder, this.repositoryUri)
-                            .fetchInstallableApplications();
-                } catch (RepositoryNotFoundException | GitAPIException e) {
-                    throw new RepositoryException(
-                            String.format("Folder '%s' is no git-repository", this.localFolder.getAbsolutePath()), e);
-                } catch (IOException e) {
-                    throw new RepositoryException("An unknown error occurred", e);
-                } finally {
-                    // close repository to free resources
-                    if (gitRepository != null) {
-                        gitRepository.close();
-                    }
+            try {
+                /*
+                 * if the repository folder previously didn't exist, clone the
+                 * repository now and checkout the correct branch
+                 */
+                if (!folderExists) {
+                    LOGGER.info("Cloning " + this);
+
+                    gitRepository = Git.cloneRepository().setURI(this.repositoryUri.toString())
+                            .setDirectory(this.localFolder)
+                            .setBranch(this.branch).call();
+                }
+                /*
+                 * otherwise open the folder and pull the newest updates from the
+                 * repository
+                 */
+                else {
+                    LOGGER.info("Opening " + this);
+
+                    gitRepository = Git.open(localFolder);
+
+                    LOGGER.info("Pulling new commits from " + this);
+
+                    gitRepository.pull().call();
                 }
+            } catch (RepositoryNotFoundException | GitAPIException e) {
+                throw new RepositoryException(
+                        String.format("Folder '%s' is no git-repository", this.localFolder.getAbsolutePath()), e);
+            } catch (IOException e) {
+                throw new RepositoryException("An unknown error occurred", e);
             } finally {
-                mutex.release();
+                // close repository to free resources
+                if (gitRepository != null) {
+                    gitRepository.close();
+                }
             }
-        } catch (InterruptedException e) {
-            throw new RepositoryException("InterruptedException occurred", e);
+        } finally {
```

Comment:
The `try { ... } finally { lock.release(); }` shouldn't be required anymore, because the calling method automatically does this in:
```java
                    try (FileLock lock = lockFileStream.getChannel().lock()) {
                        cloneOrUpdateWithLock(lock);
                    }
```